landlab/components/species_evolution/species_evolver.py

```diff
@@ -0,0 +1,673 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""Evolve life in a landscape.
+
+Component written by Nathan Lyons beginning August 2017.
+"""
+from collections import OrderedDict
+
+import numpy as np
+from pandas import DataFrame, isnull
+
+from landlab import Component
+from landlab.components.species_evolution.record import Record
+
+
+class SpeciesEvolver(Component):
+    """Evolve life in a landscape.
+
+    This component develops simulated organismal lineages of ``Taxon`` objects
+    as they evolve in response to landscape change. The history of lineages
+    (phylogeny) are tracked through the taxa. Evolution processes are conducted
+    by the Taxon objects. The taxon type, ``ZoneTaxon`` is distributed with
+    SpeciesEvolver. ``Taxon`` are intended to be subclassed for unique
+    behavior, attributes, and model approaches. The component stores Taxon
+    objects and calls their evolutionary process methods.
+
+    The general workflow to use this component in a model is
+
+    1. Instantiate the component.
+    2. Instantiate taxa.
+    3. Introduce taxa to SpeciesEvolver using the ``track_taxon`` method.
+    4. Advance the component instance in time using ``run_one_step`` method.
+
+    Taxa can be introduced at model onset and later time steps. Taxa can be
+    ``ZoneTaxon`` and other ``Taxon`` subclass types. Multiple types can be
+    tracked by the same SpeciesEvolver instance.
+
+    The spatial aspect of ``ZoneTaxon`` macroevolutionary processes is
+    determined using ``Zone`` objects. A ``ZoneController`` is used to create
+    and manage zones as well as efficiently create multiple ZoneTaxon objects.
+    See the documentation of ``ZoneController`` and ``ZoneTaxon`` for more
+    information. SpeciesEvolver knows nothing about zones and their controller,
+    meaning the concept of zones are not required for other taxon types.
+
+    Model time and other variables can be accessed with the class attribute,
+    ``record_data_frame``. Time is tracked to construct phylogeny. The unit of
+    time is not considered within the component other than the record, and can
+    be thought of as in years or whatever unit is needed. Time is advanced
+    with the ``dt`` parameter of the ``run_one_step`` method.
+
+    The geographic ranges of the taxa at the current model time are evaluated
+    during the ``run_one_step`` method. Taxa persistence or extinction is
+    determined by the taxon objects. Creation of child taxa is also determined
```

Comment:
taxa and taxon are not intuitive words to me in that I keep thinking that taxa is singular and taxon is plural. This is just me, but if you changed "child taxa" to "taxa children" it seems more clear to me.

Reply:
I relate and my intuition also switches taxon and taxa. I think I'm getting used to it.

In "Creation of child taxa is also", child is modifying taxa so flipping the words doesn't happen to fit here. I changed this part to

> Each taxon object determines if it persists or becomes extinct, and if it creates child ``Taxon`` objects.